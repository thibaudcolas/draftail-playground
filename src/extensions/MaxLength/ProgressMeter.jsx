import React, { PureComponent } from "react";

import "./ProgressMeter.css";

const getMeterColor = (progress) => {
  let color = "#1da1f2";

  if (progress >= 1) {
    color = "#ff4136";
  } else if (progress >= 0.9) {
    color = "orange";
  }

  return color;
};

/**
 * A radial progress meter, made with SVG stroke-dasharray and stroke-dashoffset.
 */
class ProgressMeter extends PureComponent {
  render() {
    const { radius, progress } = this.props;
    const diameter = radius * 2;
    const circumference = diameter * Math.PI;
    const isFull = progress >= 1;

    return (
      <svg
        height={diameter}
        width={diameter}
        className={`ProgressMeter Draftail-Icon${
          isFull ? " ProgressMeter--pulse" : ""
        }`}
      >
        <circle
          className="ProgressMeter__background"
          cx="50%"
          cy="50%"
          r={radius}
        />
        <circle
          className="ProgressMeter__progressbar"
          cx="50%"
          cy="50%"
          r={radius}
          stroke={getMeterColor(progress)}
          style={{
            strokeDasharray: circumference,
            strokeDashoffset: isFull
              ? 0
              : circumference - circumference * progress,
          }}
        />
      </svg>
    );
  }
}

export default ProgressMeter;
