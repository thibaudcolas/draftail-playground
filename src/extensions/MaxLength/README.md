# Max-length extension for Draftail

A Draftail extension to enforce a maximum length constraint on the editor content. The UI is inspired by Twitter’s character count, which highlights the text that goes over the tweets’ length limit.

The creation of this extension is available as a tutorial on [draftail.org](https://www.draftail.org/): [Extensions tutorial: max length](https://www.draftail.org/docs/extensions-tutorial-max-length)
