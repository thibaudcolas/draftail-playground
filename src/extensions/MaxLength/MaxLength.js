import { EditorState } from "draft-js";
import PropTypes from "prop-types";
import React, { Component } from "react";
import { ToolbarButton } from "draftail";
import Select from "react-simpler-select";

import Modal from "../../components/Modal";
import ProgressMeter from "./ProgressMeter";

import "./MaxLength.css";

const CONTENT_LENGTHS = {};
CONTENT_LENGTHS[140] = "Tweet";
CONTENT_LENGTHS[280] = "Double tweet";
CONTENT_LENGTHS[3 * 10 * 200] = "3-min read";
CONTENT_LENGTHS[211591 * 10 * 200] = "Crime and Punishment";

const CONTENT_LENGTH_OPTIONS = Object.keys(CONTENT_LENGTHS).map((value) => ({
  value,
  label: CONTENT_LENGTHS[value],
}));

const getDefaultThreshold = () => {
  let threshold = CONTENT_LENGTH_OPTIONS[0].value;
  try {
    const saved = JSON.parse(window.sessionStorage.getItem("threshold"));

    if (saved) {
      threshold = saved;
    }
  } catch (e) {
    console.error("sessionStorage unavailable");
  }

  return threshold;
};

/**
 * Forces a reset of the whole editor state, so text decorations are re-calculated on all blocks.
 * By default, Draft.js only updates decorations for the blocks that are under focus.
 */
const forceResetEditorState = (editorState) => {
  return EditorState.set(
    EditorState.createWithContent(
      editorState.getCurrentContent(),
      editorState.getDecorator(),
    ),
    {
      selection: editorState.getSelection(),
      undoStack: editorState.getUndoStack(),
      redoStack: editorState.getRedoStack(),
    },
  );
};

/**
 * Execute time-consuming logic out of order with performance-sensitive JS on the main thread.
 */
const delayAndIdle = (callback, timeoutHandle) => {
  if (timeoutHandle) {
    window.clearTimeout(timeoutHandle);
  }

  return window.setTimeout(() => {
    if (window.requestIdleCallback) {
      window.requestIdleCallback(callback, { timeout: 500 });
    } else {
      callback();
    }
  }, 500);
};

/**
 * A basic control showing the reading time / content length for the editorâ€™s content.
 */
class MaxLength extends Component {
  constructor(props) {
    super(props);
    const threshold = getDefaultThreshold();

    this.state = {
      isOpen: false,
      threshold,
    };

    this.onClickButton = this.onClickButton.bind(this);
    this.onRequestClose = this.onRequestClose.bind(this);
    this.onChangeThreshold = this.onChangeThreshold.bind(this);
    this.forceRenderDecorators = this.forceRenderDecorators.bind(this);
  }

  onClickButton() {
    this.setState({
      isOpen: true,
    });
  }

  onRequestClose() {
    this.setState({
      isOpen: false,
    });
  }

  forceRenderDecorators() {
    const { getEditorState, onChange } = this.props;
    const editorState = getEditorState();

    this.timeoutHandle = delayAndIdle(() => {
      onChange(forceResetEditorState(editorState));
    }, this.timeoutHandle);
  }

  onChangeThreshold(value) {
    const threshold = Number(value);
    this.setState({
      threshold,
    });

    if (window.sessionStorage) {
      try {
        window.sessionStorage.setItem("threshold", JSON.stringify(threshold));
      } catch (e) {
        console.error("sessionStorage unavailable");
      }
    }

    this.forceRenderDecorators();
  }

  componentDidUpdate() {
    const { getEditorState } = this.props;
    const editorState = getEditorState();
    const lastChange = editorState.getLastChangeType();

    if (lastChange) {
      this.forceRenderDecorators();
    }
  }

  render() {
    const { getEditorState } = this.props;
    const { isOpen, threshold } = this.state;
    const editorState = getEditorState();
    const content = editorState.getCurrentContent();
    const contentLength = content
      .getBlockMap()
      .reduce((length, block) => length + block.getLength(), 0);

    return (
      <React.Fragment>
        <ToolbarButton
          name="MAX_LENGTH"
          title={`Length: ${CONTENT_LENGTHS[threshold]}`}
          icon={
            <ProgressMeter radius={8} progress={contentLength / threshold} />
          }
          onClick={this.onClickButton}
        />
        <Modal
          onRequestClose={this.onRequestClose}
          onAfterOpen={this.onAfterOpen}
          isOpen={isOpen}
          contentLabel="Maximum length"
        >
          <label>
            <p>How long should the content be?</p>
            <Select
              className="MaxLength__select"
              value={threshold}
              options={CONTENT_LENGTH_OPTIONS}
              onChange={this.onChangeThreshold}
            />
          </label>
        </Modal>
      </React.Fragment>
    );
  }
}

MaxLength.propTypes = {
  getEditorState: PropTypes.func.isRequired,
};

/**
 * Creates text decorations based on the length of text of each block, and its preceding blocks.
 */
export class MaxLengthDecorator {
  constructor() {
    this.component = this.renderDecoration.bind(this);
    this.strategy = this.createDecorations.bind(this);
  }

  renderDecoration({ children }) {
    return <mark className="overflow-mark">{children}</mark>;
  }

  createDecorations(block, callback, contentState) {
    const blockKey = block.getKey();
    const blockLength = block.getLength();

    const previousContentLength = contentState
      .getBlockMap()
      .takeUntil((block) => block.getKey() === blockKey)
      .reduce((length, block) => length + block.getLength(), 0);

    const currentLength = previousContentLength + blockLength;
    const threshold = getDefaultThreshold();
    const shouldDecorate = currentLength > threshold;

    if (shouldDecorate) {
      // Decorate from the first character going over the threshold, to the end of the block.
      const startOffset = Math.max(0, threshold - previousContentLength);
      callback(startOffset, blockLength);
    }
  }
}

export default MaxLength;
