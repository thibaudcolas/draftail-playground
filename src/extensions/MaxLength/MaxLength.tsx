import { ContentState, EditorState } from "draft-js";
import React, { Component } from "react";
import { ToolbarButton, ControlProps } from "draftail";
import Select from "react-simpler-select";

import Modal from "../../components/Modal";
import ProgressMeter from "./ProgressMeter";

import "./MaxLength.css";

const CONTENT_LENGTHS = {
  "140": "Tweet",
  "280": "Double tweet",
  // 3 * 10 * 200
  "6000": "3-min read",
  // 211591 * 10 * 200
  "423182000": "Crime and Punishment",
} as const;

const CONTENT_LENGTH_OPTIONS = Object.entries(CONTENT_LENGTHS).map(
  ([value, label]) => ({
    value,
    label,
  }),
);

const getDefaultThreshold = () => {
  let threshold = Number(CONTENT_LENGTH_OPTIONS[0].value);
  try {
    const raw = window.sessionStorage.getItem("threshold");
    if (raw) {
      const saved = JSON.parse(raw);

      if (saved) {
        threshold = Number(saved);
      }
    }
  } catch (e) {
    console.error("sessionStorage unavailable");
  }

  return threshold;
};

/**
 * Forces a reset of the whole editor state, so text decorations are re-calculated on all blocks.
 * By default, Draft.js only updates decorations for the blocks that are under focus.
 */
const forceResetEditorState = (editorState: EditorState) => {
  return EditorState.set(
    EditorState.createWithContent(
      editorState.getCurrentContent(),
      editorState.getDecorator(),
    ),
    {
      selection: editorState.getSelection(),
      undoStack: editorState.getUndoStack(),
      redoStack: editorState.getRedoStack(),
    },
  );
};

type RequestIdleCallbackHandle = any;
type RequestIdleCallbackOptions = {
  timeout: number;
};
type RequestIdleCallbackDeadline = {
  readonly didTimeout: boolean;
  timeRemaining: () => number;
};

declare global {
  interface Window {
    requestIdleCallback: (
      callback: (deadline: RequestIdleCallbackDeadline) => void,
      opts?: RequestIdleCallbackOptions,
    ) => RequestIdleCallbackHandle;
    cancelIdleCallback: (handle: RequestIdleCallbackHandle) => void;
  }
}

/**
 * Execute time-consuming logic out of order with performance-sensitive JS on the main thread.
 */
const delayAndIdle = (callback: () => void, timeoutHandle: number) => {
  if (timeoutHandle) {
    window.clearTimeout(timeoutHandle);
  }

  return window.setTimeout(() => {
    if (window.requestIdleCallback) {
      window.requestIdleCallback(callback, { timeout: 500 });
    } else {
      callback();
    }
  }, 500);
};

/**
 * A basic control showing the reading time / content length for the editorâ€™s content.
 */
class MaxLength extends Component<
  ControlProps,
  { isOpen: boolean; threshold: keyof typeof CONTENT_LENGTHS }
> {
  private timeoutHandle: number;

  constructor(props: ControlProps) {
    super(props);
    const threshold = getDefaultThreshold();

    this.state = {
      isOpen: false,
      threshold,
    };

    this.timeoutHandle = 0;

    this.onClickButton = this.onClickButton.bind(this);
    this.onRequestClose = this.onRequestClose.bind(this);
    this.onChangeThreshold = this.onChangeThreshold.bind(this);
    this.forceRenderDecorators = this.forceRenderDecorators.bind(this);
  }

  onClickButton() {
    this.setState({
      isOpen: true,
    });
  }

  onRequestClose() {
    this.setState({
      isOpen: false,
    });
  }

  forceRenderDecorators() {
    const { getEditorState, onChange } = this.props;
    const editorState = getEditorState();

    this.timeoutHandle = delayAndIdle(() => {
      onChange(forceResetEditorState(editorState));
    }, this.timeoutHandle);
  }

  onChangeThreshold(value: keyof typeof CONTENT_LENGTHS) {
    const threshold = value;
    this.setState({
      threshold,
    });

    if (window.sessionStorage) {
      try {
        window.sessionStorage.setItem("threshold", JSON.stringify(threshold));
      } catch (e) {
        console.error("sessionStorage unavailable");
      }
    }

    this.forceRenderDecorators();
  }

  componentDidUpdate() {
    const { getEditorState } = this.props;
    const editorState = getEditorState();
    const lastChange = editorState.getLastChangeType();

    if (lastChange) {
      this.forceRenderDecorators();
    }
  }

  render() {
    const { getEditorState } = this.props;
    const { isOpen, threshold } = this.state;
    const editorState = getEditorState();
    const content = editorState.getCurrentContent();
    const contentLength = content.getBlockMap().reduce(
      // @ts-ignore
      (length: number, block: { getLength: () => number }): number =>
        length + block.getLength(),
      0,
    );

    return (
      <React.Fragment>
        <ToolbarButton
          name="MAX_LENGTH"
          title={`Length: ${CONTENT_LENGTHS[threshold]}`}
          icon={
            <ProgressMeter
              radius={8}
              progress={contentLength / Number(threshold)}
            />
          }
          onClick={this.onClickButton}
        />
        <Modal
          onRequestClose={this.onRequestClose}
          onAfterOpen={() => {}}
          isOpen={isOpen}
          contentLabel="Maximum length"
        >
          <label>
            <p>How long should the content be?</p>
            <Select
              className="MaxLength__select"
              value={threshold}
              options={CONTENT_LENGTH_OPTIONS}
              onChange={this.onChangeThreshold}
            />
          </label>
        </Modal>
      </React.Fragment>
    );
  }
}

/**
 * Creates text decorations based on the length of text of each block, and its preceding blocks.
 */
export class MaxLengthDecorator {
  private component: (props: { children: React.ReactNode }) => JSX.Element;

  constructor() {
    this.component = this.renderDecoration.bind(this);
    // @ts-ignore
    this.strategy = this.createDecorations.bind(this);
  }

  renderDecoration({ children }: { children: React.ReactNode }) {
    return <mark className="overflow-mark">{children}</mark>;
  }

  createDecorations(
    // @ts-ignore
    block,
    callback: (start: number, length: number) => void,
    contentState: ContentState,
  ) {
    const blockKey = block.getKey();
    const blockLength = block.getLength();

    const previousContentLength = contentState
      .getBlockMap()
      // @ts-ignore
      .takeUntil((block) => block.getKey() === blockKey)
      // @ts-ignore
      .reduce((length, block) => length + block.getLength(), 0);

    const currentLength = previousContentLength + blockLength;
    const threshold = getDefaultThreshold();
    const shouldDecorate = currentLength > threshold;

    if (shouldDecorate) {
      // Decorate from the first character going over the threshold, to the end of the block.
      const startOffset = Math.max(0, threshold - previousContentLength);
      callback(startOffset, blockLength);
    }
  }
}

export default MaxLength;
