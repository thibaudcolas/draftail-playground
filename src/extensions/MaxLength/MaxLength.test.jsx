import React from "react";
import { convertFromRaw, EditorState } from "draft-js";
import { shallow, mount } from "enzyme";
import MaxLength, { MaxLengthDecorator } from "./MaxLength";

describe("MaxLength", () => {
  it("works", () => {
    expect(
      shallow(<MaxLength getEditorState={() => EditorState.createEmpty()} />),
    ).toMatchSnapshot();
  });

  it("recovers from sessionStorage / JSON parsing issues", () => {
    window.sessionStorage.setItem('threshold', "140")
    jest.spyOn(JSON, "parse").mockImplementationOnce(() => {
      throw new Error();
    });

    expect(() => {
      mount(<MaxLength getEditorState={() => EditorState.createEmpty()} />);
    }).toThrow();

    jest.restoreAllMocks();
  });
});

describe("MaxLengthDecorator", () => {
  it("decorates", () => {
    const decorator = new MaxLengthDecorator();
    expect(shallow(decorator.component(<div>Test!</div>)))
      .toMatchInlineSnapshot(`
<mark
  className="overflow-mark"
/>
`);
  });

  describe("finds decorations", () => {
    it("single block below threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test",
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });

    it("single block above threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test".repeat(200),
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });

    it("multiple blocks below threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test",
          },
          {
            key: "b",
            text: "test",
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });

    it("multiple blocks above threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test".repeat(200),
          },
          {
            key: "b",
            text: "test",
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });
  });
});
