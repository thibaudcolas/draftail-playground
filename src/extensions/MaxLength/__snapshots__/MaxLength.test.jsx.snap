// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxLength works 1`] = `
<Fragment>
  <ToolbarButton
    icon={
      <ProgressMeter
        progress={0}
        radius={8}
      />
    }
    name="MAX_LENGTH"
    onClick={[Function]}
    title="Length: Tweet"
  />
  <Modal
    contentLabel="Maximum length"
    isOpen={false}
    onAfterOpen={[Function]}
    onRequestClose={[Function]}
  >
    <label>
      <p>
        How long should the content be?
      </p>
      <Select
        className="MaxLength__select"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Tweet",
              "value": "140",
            },
            Object {
              "label": "Double tweet",
              "value": "280",
            },
            Object {
              "label": "3-min read",
              "value": "6000",
            },
            Object {
              "label": "Crime and Punishment",
              "value": "423182000",
            },
          ]
        }
        placeholder={null}
        value={140}
      />
    </label>
  </Modal>
</Fragment>
`;
